Finality RPC

JoshOrndorff opened the pull request:
Solves https://purestake.atlassian.net/browse/MOON-911

This PR adds two new RPC endpoints to the Moonbeam node, that are useful for checking whether an on-chain event is finalized.

- [x] `moon_isBlockFinalized`
- [ ] `moon_isTxFinalized`

These methods accept either Substrate-style or etheruem-style hashes as their input.

These methods are useful for high-stakes projects that require a strong notion of finality, such as bridges. Typically, in Ethereum mainnet, such applications will wait for a certain number of block confirmations which guarantees a certain degree of finality according to the PoW mechanism. However, in Moonbeam, we do not have PoW and we do have deterministic finality provided by the relay chain's Grandpa validators. These RPC endpoints provide a simple way for applications to adapt to moonbeam with minimal changes to their code.

This PR still needs tests. I've thought of 10 cases:
- [ ] Substrate {tx, block} is finalized
- [ ] Substrate {tx, block} is not finalized
- [ ] Eth {tx, block} is finalized
- [ ] Eth {tx, block} is not finalized
- [ ] {Tx, Block} hash is not found in database

notlesh commented on client/rpc/finality/src/lib.rs:
```diff
@@ -0,0 +1,124 @@
+	let query_height = client
+		.number(substrate_hash)
+		.expect("Blocks in best chain should have number")
+		.expect("uhhhh");
```
:eyes:

JoshOrndorff replied:
Yeah, I wasn't sure why there were nested results. Maybe this is the block-not-in-database case? If so, then I should probably use a `?` rather than `expect`ing the block to be present.

JoshOrndorff replied:
Okay, I figured it out and updated the `expect` messages in 7b82dd8

notlesh commented on client/rpc/finality/src/lib.rs:
```diff
@@ -0,0 +1,124 @@
+		.expect("uhhhh");
+	let finalized_height = client.info().finalized_number;
+
+	query_height <= finalized_height
```
Does this need to take forks into account? E.g. the requested block/txn should be an ancestor of the finalized block?

notlesh replied:
Oh, you ensure `is_canon()` above...

JoshOrndorff replied:
Yeah, that's the key idea. Once we know the block is currently canon, we just have to determine whether it is in the finalized prefix or not.

JoshOrndorff commented:
I've updated this PR with master, addressed the feedback given so far, and clarified my thoughts in the code comments. This is now ready to be taken over by another developer. @tgmichel and @librelois have both expressed interest.

librelois commented:
Replaced by #1184
